File: run_migration.py
#!/usr/bin/env python3
"""
Database migration script to add missing columns to users table
This requires postgres superuser credentials
"""

import psycopg2
from psycopg2 import sql
import sys

def run_migration():
    # Try connecting with postgres superuser
    credentials = [
        {'user': 'postgres', 'password': 'postgres', 'host': 'localhost', 'port': 5432},
        {'user': 'postgres', 'password': 'password', 'host': 'localhost', 'port': 5432},
        {'user': 'postgres', 'password': 'admin', 'host': 'localhost', 'port': 5432},
        {'user': 'postgres', 'password': '', 'host': 'localhost', 'port': 5432},  # No password
    ]
    
    connection = None
    for creds in credentials:
        try:
            print(f"Attempting to connect with user: {creds['user']}")
            connection = psycopg2.connect(
                user=creds['user'],
                password=creds['password'],
                host=creds['host'],
                port=creds['port'],
                database='rasa_db'
            )
            print("✅ Connected successfully!")
            break
        except psycopg2.OperationalError as e:
            print(f"❌ Failed: {e}")
            continue
    
    if not connection:
        print("\n❌ Could not connect with any credentials")
        print("Please run the migration manually with:")
        print("  psql -U postgres -d rasa_db -f migrate_add_columns.sql")
        return False
    
    try:
        cursor = connection.cursor()
        
        # List of columns to add with their definitions
        columns = [
            ('roll_number', 'VARCHAR(100)'),
            ('phone', 'VARCHAR(20)'),
            ('location', 'VARCHAR(255)'),
            ('department', 'VARCHAR(255)'),
            ('bio', 'TEXT'),
            ('github', 'VARCHAR(255)'),
            ('portfolio', 'VARCHAR(255)'),
            ('batch', 'VARCHAR(50)'),
            ('skills', 'VARCHAR(500)'),
            ('qualifications', 'TEXT'),
            ('research_interests', 'TEXT'),
            ('office_hours', 'VARCHAR(255)'),
            ('current_company', 'VARCHAR(255)'),
            ('job_title', 'VARCHAR(255)'),
            ('linkedin', 'VARCHAR(255)'),
            ('open_to_referrals', 'BOOLEAN DEFAULT FALSE'),
            ('is_alumni', 'BOOLEAN DEFAULT FALSE'),
            ('updated_at', 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP'),
        ]
        
        # Add each column
        for column_name, column_type in columns:
            try:
                alter_sql = f"ALTER TABLE users ADD COLUMN IF NOT EXISTS {column_name} {column_type}"
                cursor.execute(alter_sql)
                print(f"✅ Added/verified column: {column_name}")
            except psycopg2.Error as e:
                print(f"⚠️  Column {column_name}: {e.pgerror}")
        
        # Grant privileges to rasa_user
        print("\nGranting privileges to rasa_user...")
        cursor.execute("GRANT ALL PRIVILEGES ON TABLE users TO rasa_user")
        cursor.execute("GRANT ALL PRIVILEGES ON SEQUENCE users_id_seq TO rasa_user")
        print("✅ Privileges granted to rasa_user")
        
        # Verify the schema
        print("\n📋 Current users table schema:")
        cursor.execute("""
            SELECT column_name, data_type, is_nullable
            FROM information_schema.columns
            WHERE table_name = 'users'
            ORDER BY ordinal_position
        """)
        
        for row in cursor.fetchall():
            print(f"  {row[0]:30} {row[1]:20} nullable={row[2]}")
        
        connection.commit()
        print("\n✅ Migration completed successfully!")
        return True
        
    except psycopg2.Error as e:
        print(f"\n❌ Migration failed: {e}")
        connection.rollback()
        return False
    finally:
        cursor.close()
        connection.close()

if __name__ == '__main__':
    success = run_migration()
    sys.exit(0 if success else 1)

File: run_migration.js
/**
 * Database Migration Runner
 * Executes migration.sql to update schema for roll_number support
 * 
 * Run: node run_migration.js
 */

const fs = require('fs');
const path = require('path');
const { Pool } = require('pg');

const pool = new Pool({
    user: 'rasa_user',
    host: 'localhost',
    database: 'rasa_db',
    password: 'rootadmin',
    port: 5432,
});

async function runMigration() {
    const client = await pool.connect();
    
    try {
        // Read migration SQL file
        const migrationPath = path.join(__dirname, 'migration.sql');
        const sql = fs.readFileSync(migrationPath, 'utf8');
        
        console.log('🔄 Running database migration...\n');
        
        // Split by semicolons and execute each statement
        const statements = sql
            .split(';')
            .map(stmt => stmt.trim())
            .filter(stmt => stmt.length > 0 && !stmt.startsWith('--'));
        
        for (const statement of statements) {
            try {
                console.log(`Executing: ${statement.substring(0, 50)}...`);
                await client.query(statement);
                console.log('✅ Success\n');
            } catch (err) {
                // Some errors like "already exists" are OK
                if (err.code === '42P07' || err.message.includes('already exists')) {
                    console.log('⚠️  Already exists (skipped)\n');
                } else if (err.message.includes('does not exist')) {
                    console.log('⚠️  Does not exist (skipped)\n');
                } else {
                    throw err;
                }
            }
        }
        
        console.log('✨ Migration completed successfully!\n');
        
        // Verify schema
        console.log('📊 Verifying schema changes...\n');
        
        const columns = await client.query(`
            SELECT column_name, data_type
            FROM information_schema.columns 
            WHERE table_name='users'
            ORDER BY ordinal_position
        `);
        
        console.log('Current users table columns:');
        columns.rows.forEach(col => {
            console.log(`  ✓ ${col.column_name} (${col.data_type})`);
        });
        
        // Check if roll_number column exists
        const hasRollNumber = columns.rows.some(col => col.column_name === 'roll_number');
        if (hasRollNumber) {
            console.log('\n✅ roll_number column is now available');
        }
        
        // Show constraints
        const constraints = await client.query(`
            SELECT constraint_name
            FROM information_schema.table_constraints
            WHERE table_name='users' AND constraint_type != 'PRIMARY KEY'
        `);
        
        console.log('\nActive constraints:');
        constraints.rows.forEach(cons => {
            console.log(`  ✓ ${cons.constraint_name}`);
        });
        
    } catch (err) {
        console.error('❌ Migration failed:', err.message);
        process.exit(1);
    } finally {
        client.release();
        await pool.end();
    }
}

runMigration();
